integration-tests/java/src/integration-test/java/com/stellarstation/api/test/satellite/SatelliteStreamerTest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019 Infostellar, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.stellarstation.api.test.satellite;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.protobuf.ByteString;
+import com.stellarstation.api.test.auth.ApiClientModule;
+import com.stellarstation.api.v1.SatelliteStreamRequest;
+import com.stellarstation.api.v1.SatelliteStreamResponse;
+import com.stellarstation.api.v1.SendSatelliteCommandsRequest;
+import com.stellarstation.api.v1.monitoring.AntennaState;
+import com.stellarstation.api.v1.monitoring.GroundStationState;
+import dagger.Component;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Singleton;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class SatelliteStreamerTest {
+  private SatelliteStreamer streamer;
+
+  private static final String SATELLITE_ID = "98";
+
+  @BeforeEach
+  void setUp() {
+    streamer = DaggerSatelliteStreamerTest_SatelliteStreamerComponent.create().streamer();
+  }
+
+  @Component(modules = ApiClientModule.class)
+  @Singleton
+  interface SatelliteStreamerComponent {
+    SatelliteStreamer streamer();
+  }
+
+  @Test
+  void openStream() throws InterruptedException {
+    final CountDownLatch latch = new CountDownLatch(1);
+    final BlockingQueue<SendSatelliteCommandsRequest> queue =
```

Comment:
It might be cleaner to restructure this a bit. It's kind of strange to have both the latch and the blocking queue.

My thought is to get rid of both the latch and the queue.

You could use a SettableFuture to store the final result of the second command. You can use Futures.withTimeout to wait for that to be ready on the main thread. Inside the onError you can set an exception on the future in the onError case (onComplete shouldn't do anything).

I'm assuming you can directly call requestObserver.onNext from inside the onNext callback rather than adding to a queue.

Reply:
As we discussed, you can't access requestObserver directly in onNext because it hasn't been assigned yet.

However, there is a kind of workaround. You can use a SettableFuture to hold the requestObserver and access the future inside onNext.

SettableFuture<StreamObserver<SatelliteStreamRequest>> requestObserverFuture = SettableFuture.create();
... requestObserver = streamer.openStream(
    .... onNext() ....
            Futures.getUnchecked(requestObserverFuture).onNext( .... )
....
requestObserverFuture.set(requestObserver);

Do you think it will work?

Reply:
Thanks, I used the idea. It is much cleaner that we can call the requestObserer in the onNext in ReponseObserver.